app/lib/multistep/controller.rb

```diff
@@ -0,0 +1,137 @@
+# frozen_string_literal: true
+
+module Multistep
+  module Controller
+    extend ActiveSupport::Concern
+
+    class_methods do
+      def form(value = nil, as: nil) # rubocop:disable Naming/MethodParameterName
+        if value
+          @form = value
+          if as
+            alias_method as, :form
+            helper_method as
+          end
+        else
+          @form
+        end
+      end
+
+      def session_key(value = nil)
+        if value
+          @session_key = value
+        else
+          @session_key ||= @form && @form.model_name.param_key
+        end
+      end
+
+      def params_key(value = nil)
+        if value
+          @params_key = value
+        else
+          @params_key ||= @form && @form.model_name.param_key
+        end
+      end
+
+      def escape_route(&block)
+        if block
+          @escape_route = block
+        else
+          @escape_route || raise("Escape route not defined!")
+        end
+      end
+
+      def result(as: name) # rubocop:disable Naming/MethodParameterName
+        alias_method as, :result
+        helper_method as
+      end
+    end
+
+    included do
+      helper_method :back_link_path
+      before_action :ensure_form_present, only: %i[show update complete]
+      after_action :remove_form, only: :complete
+    end
+
+    attr_reader :result
+
+    def start
+      reset_form
+      redirect_to action: :show, step: form.next_step
+    end
+
+    def show
+      render current_step
+    end
+
+    def update
+      form.assign_attributes(form_params)
+
+      if form.valid?(current_step)
+        form.record_completed_step current_step
+        store_form_in_session
+        redirect_to action: :show, step: step_param(form.next_step)
+      else
+        render current_step
+      end
+    end
+
+    def complete
+      @result = add_participant_form.save!
+    end
+
+  private
+
+    def reset_form
+      session.delete(self.class.session_key)
+      form = self.class.form.new
+      setup_form(form)
+      @form = form
+      store_form_in_session
+    end
+
+    def form
+      @form ||= self.class.form.new(session[self.class.session_key])
+    end
+
+    def store_form_in_session
+      session[self.class.session_key] = form.attributes
+    end
+
+    def current_step
+      params[:step].underscore.to_sym
+    end
+
+    def setup_form(form)
+      # noop
+    end
+
+    def back_link_path
+      previous_step = add_participant_form.previous_step(from: current_step)
+      return escape_route unless previous_step
+
+      { action: :show, step: step_param(previous_step) }
+    end
+
+    def form_params
+      attributes = form.class.steps[current_step].attributes
+      params[self.class.params_key]&.permit(*attributes) || {}
+    end
+
+    def ensure_form_present
+      redirect_to escape_route unless session.key?(self.class.session_key)
+    end
+
+    def escape_route
+      instance_exec(&self.class.escape_route)
+    end
```

Comment:
We already have this method in `public` scope - except there it takes an argument. Is that intentional? Also, what is `instance_exec` exactly?

Reply:
So yes, there are two `escape_routes` but in two different context. The first exists on the class, forming part of the public API to configure the behaviour of the controller. The other call is defined on the instance of the controller and executes the block given in the configuration in the context of the controller instance. The reason for this to always be a block is that neither url_helpers nor url_options are not available in the class context. Additionally, the back link can be dynamic (and there's already another PR that makes it dynamic for the add_participant journey.

`instance_exec` is an awesome method defined on `Object` used pretty much for meta-programming only - it executes given block in the context of the receiver (so changes the `self` inside the block), giving the block full access to the object internals, including access to its instance variables. As always, it needs to be used sparingly and only in certain scenarios.  I think it is applicable here as the block is to be defined within the same class, so it doesn't violate object encapsulation.

There's another similar method `instance_eval` which works very similarly with two important differences:
1) when executed with lambda instances passed as a block, `instance_eval` expects lambda to have arity of 1(yields object), while `instance_exec` expects arity of 0.
2). `instance_exec` additionally allows yielding more data into the block.

In our case, we're capturing the block resulting in a proc object, so 1) is not an issue, we also do not use 2) so we could use both. I went for `instance_exec` as I see no point of optional yield of self into the block.

Other options here would be:
1) Provide an overridable method. This however expects developer to know which method to override on the instance and requires the abstract, default implementation (in this case noop or to raise an exception). I used this pattern for `setup_form` but I think I should probably use declarative form there as well.
2) instead of instance_exec, we could do `yield self`. This would require declaration to be: escape_route { |controller| controller.url_helper }` which is a bit cumbersome (especially for more complicated logic) given that pretty much every single call would be prefixed with `controller.` and we'd loose access to private methods and variables (sort-of, we can always do `send` :P)